### PR TITLE
Fix FakeTensorMode handling of aten.embedding.default for meta tensor indices (supports torch.compile with meta inputs)

### LINKED ISSUE #166644: [Dynamo] Device propagation failure in aten.embedding.default with mixed cpu/meta devices
### Description
PyTorch Dynamo compilation fails when processing `aten.embedding.default` operations that involve tensors on different devices (`cpu` and `meta`). The error occurs during FakeTensor propagation in the compilation phase, while normal eager execution works correctly.
code:
'''
import torch
import torch.nn as nn
import torch.nn.functional as F

class DotProductAttention(nn.Module):

    def __init__(self, embed_size, heads):
        super(DotProductAttention, self).__init__()
        self.embed_size = embed_size
        self.heads = heads
        self.head_dim = embed_size // heads
        assert self.head_dim * heads == embed_size
        self.values = nn.Linear(embed_size, embed_size)
        self.keys = nn.Linear(embed_size, embed_size)
        self.queries = nn.Linear(embed_size, embed_size)
        self.fc_out = nn.Linear(embed_size, embed_size)

    def forward(self, values, keys, queries, mask=None):
        N = queries.shape[0]
        (value_len, key_len, query_len) = (values.shape[1], keys.shape[1], queries.shape[1])
        values = self.values(values)
        keys = self.keys(keys)
        queries = self.queries(queries)
        values = values.reshape(N, value_len, self.heads, self.head_dim)
        keys = keys.reshape(N, key_len, self.heads, self.head_dim)
        queries = queries.reshape(N, query_len, self.heads, self.head_dim)
        energy = torch.einsum('nqhd,nkhd->nhqk', [queries, keys])
        if mask is not None:
            energy = energy.masked_fill(mask == 0, float('-1e20'))
        attention = torch.softmax(energy / self.embed_size ** (1 / 2), dim=3)
        out = torch.einsum('nhql,nlhd->nqhd', [attention, values])
        out = out.reshape(N, query_len, self.heads * self.head_dim)
        out = self.fc_out(out)
        return out

class DotProductAttentionBlock(nn.Module):

    def __init__(self, embed_size, heads, forward_expansion, dropout):
        super(DotProductAttentionBlock, self).__init__()
        self.attention = DotProductAttention(embed_size, heads)
        self.norm1 = nn.LayerNorm(embed_size)
        self.norm2 = nn.LayerNorm(embed_size)
        self.feed_forward = nn.Sequential(nn.Linear(embed_size, forward_expansion * embed_size), nn.ReLU(), nn.Linear(forward_expansion * embed_size, embed_size))
        self.dropout = nn.Dropout(dropout)

    def forward(self, value, key, query, mask=None):
        attention = self.attention(value, key, query, mask)
        x = self.dropout(self.norm1(attention + query))
        forward = self.feed_forward(x)
        out = self.dropout(self.norm2(forward + x))
        return out

class DotProductAttentionDNN(nn.Module):

    def __init__(self, input_size, embed_size, num_layers, heads, forward_expansion, output_size, dropout, max_length):
        super(DotProductAttentionDNN, self).__init__()
        self.embed_size = embed_size
        self.word_embedding = nn.Embedding(input_size, embed_size)
        self.position_embedding = nn.Embedding(max_length, embed_size)
        self.layers = nn.ModuleList([DotProductAttentionBlock(embed_size, heads, forward_expansion, dropout) for _ in range(num_layers)])
        self.fc = nn.Linear(embed_size, output_size)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x, mask=None):
        (N, seq_length) = x.shape
        positions = torch.arange(0, seq_length).expand(N, seq_length).to('meta')
        out = self.dropout(self.word_embedding(x) + self.position_embedding(positions))
        for layer in self.layers:
            out = layer(out, out, out, mask)
        out = self.fc(out)
        return out

def get_default_model():
    return DotProductAttentionDNN(input_size=1000, embed_size=128, num_layers=2, heads=4, forward_expansion=2, output_size=10, dropout=0.1, max_length=20)

def get_sample_inputs(batch_size=2, seq_length=10):
    x = torch.randint(0, 1000, (batch_size, seq_length))
    mask = torch.ones(batch_size, 1, 1, seq_length)
    return (x, mask)

def run_inference(model, x, mask):
    with torch.no_grad():
        output = model(x, mask)
    return output

def main():
    model = get_default_model()
    model.eval()
    (x, mask) = get_sample_inputs()

    output_normal = run_inference(model, x, mask)
    print('Normal inference successful')
    print(f'Output shape: {output_normal.shape}')

    try:
        compiled_model = torch.compile(model)
        compiled_model.eval()
        output_compiled = run_inference(compiled_model, x, mask)
        print('Compiled inference successful')
        print(f'Compiled output shape: {output_compiled.shape}')

        diff = torch.abs(output_normal - output_compiled).max().item()
        print(f'Output difference: {diff:.6f}')

    except Exception as exception:
        print("Compilation failed:", exception)

    print(f'Model parameters: {sum((p.numel() for p in model.parameters()))}')

if __name__ == '__main__':
    main()
'''
output:
'''
Normal inference successful
Output shape: torch.Size([2, 10, 10])
Compilation failed: Dynamo failed to run FX node with fake tensors: call_function <function embedding at 0x7fd452657dc0>(*(FakeTensor(..., device='meta', size=(2, 10), dtype=torch.int64), Parameter(FakeTensor(..., size=(20, 128), requires_grad=True)), None, None, 2.0, False, False), **{}): got RuntimeError('Unhandled FakeTensor Device Propagation for aten.embedding.default, found two different devices cpu, meta')

from user code:
   File line 69, in forward
    out = self.dropout(self.word_embedding(x) + self.position_embedding(positions))
 line 190, in forward
    return F.embedding(

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
'''

### Versions

PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.5.0-6ubuntu2) 9.5.0
Clang version: Could not collect
CMake version: version 4.0.3
Libc version: glibc-2.39

Python version: 3.9.7 (default, Jul 16 2025, 16:34:47) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-29-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 580.65.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 39 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: GenuineIntel
Model name: Intel(R) Core(TM) i9-14900HX
CPU family: 6
Model: 183
Thread(s) per core: 2
Core(s) per socket: 24
Socket(s): 1
Stepping: 1
CPU(s) scaling MHz: 31%
CPU max MHz: 5800.0000
CPU min MHz: 800.0000
BogoMIPS: 4838.40
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization: VT-x
L1d cache: 896 KiB (24 instances)
L1i cache: 1.3 MiB (24 instances)
L2 cache: 32 MiB (12 instances)
L3 cache: 36 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Ghostwrite: Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] nvidia-cublas-cu11==11.10.3.66
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu11==11.7.101
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu11==11.7.99
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu11==11.7.99
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu11==8.5.0.96
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu11==10.2.10.91
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu11==11.4.0.1
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu11==11.7.4.91
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu11==2.14.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu11==11.7.91
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.1
[pip3] torchdiffeq==0.2.5
[pip3] torchvision==0.22.1
[pip3] triton==3.3.1
[conda] Could not collect

### PR DESCRIPTION
**Pull request by adam10cole**
Fixes #166644

This PR fixes FakeTensorMode handling of `aten.embedding.default` when the input indices are meta tensors. Without this fix, `torch.compile(...,backend="eager")` fails with:

    Unhandled FakeTensor Device Propagation for aten.embedding.default,
    found two different devices cpu, meta

This occurs during Dynamo’s FakeTensor-based shape propagation phase.

## Fix/Changes 
A FakeTensorMode operator implementation is added for `aten.embedding.default` which mirrors how the project already handled FakeTensor for `embedding_bag`. It calls the meta kernel (`torch._meta_registrations.embedding`) under `FakeTensorMode`. Essentially, it just ensures these things:

- correct output shape/dtype inference via meta kernels
- no device propagation errors (mixed cpu/meta is allowed)
- no fallback to real kernels
- the output is a FakeTensor with proper fake_device tracking

Also, within `_find_common_device` I created a special case for `aten.where` to avoid a related mixed-device assertion.

## Tests
Added a full FakeTensor test file (`test/test_fake_embedding.py`) that covers:

1. `nn.Embedding(x_meta)` inside FakeTensorMode
2. `torch.compile(..., backend="eager")` with meta tensor indices
3. shape/dtype-only behavior (`propagate_real_tensors=False`)
4. multi-dimensional meta indices
5. non-contiguous meta indices
6. padding_idx behavior
7. two embeddings in the same model

These tests fail on the current main branch but pass with this patch.

## Motivation
This fixes the issue: https://github.com/pytorch/pytorch/issues/166644 and enables models using meta tensor shapes + embedding to work with `torch.compile` and FakeTensorMode without errors.

## Backward compatibility 
FakeTensorMode only affects internal tracing/meta-tracing paths and has no user-visible behavior change outside FakeTensorMode so this should be BC safe.

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
Figure out what the right thing to do is
